### PR TITLE
[PT FE] Add aten::fft_rfftn and aten::fft_irfft complex replacers

### PR DESCRIPTION
**Pull request by mmikolajcz**
### Details:
 - *Add replacers for patterns:*
     - `aten::fft_rfftn -> {aten::real, aten::imag}`
     - `aten::complex -> aten::fft_irfftn`
 - *Add test cases*
 - *Fix issue with fft_base.cpp not accepting dynamic types

### Tickets:
 - *110806*
 - *112881*

### CODE REVIEW

File: src/frontends/pytorch/src/utils.cpp [hunk 1]
```diff
@@ -307,6 +307,16 @@ std::shared_ptr<ov::op::util::FrameworkNode> cast_fw_node(std::shared_ptr<Node>
     return fw_node;
 }
 
+bool is_none_node(std::shared_ptr<Node> node) {
```
Comment (mvafin) [inline src/frontends/pytorch/src/utils.cpp:310]: You use this function on output, so better to change signature to `bool is_none_node(Output<Node> node)`

File: src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        auto const_scalar_0 = v0::Constant::create(element::i32, Shape{}, {0});
+        auto const_1 = v0::Constant::create(element::i32, Shape{1}, {1});
+        auto const_scalar_1 = v0::Constant::create(element::i32, Shape{}, {1});
+        auto const_2 = v0::Constant::create(element::i32, Shape{1}, {2});
+        auto fw_node_complex_input = cast_fw_node(irfftn_op->input_value(0).get_node_shared_ptr(), "aten::complex");
+        if (!fw_node_complex_input) {
+            return false;
+        }
+        auto real = std::make_shared<v0::Unsqueeze>(fw_node_complex_input->input(0).get_source_output(), const_neg_1);
+        auto imag = std::make_shared<v0::Unsqueeze>(fw_node_complex_input->input(1).get_source_output(), const_neg_1);
```
Comment (mvafin) [inline src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp:58]:
```suggestion
        auto real = std::make_shared<v0::Unsqueeze>(fw_node_complex_input->input_value(0), const_neg_1);
        auto imag = std::make_shared<v0::Unsqueeze>(fw_node_complex_input->input_value(1), const_neg_1);
```

File: src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        NodeVector complex = {real, imag};
+        auto input = std::make_shared<v0::Concat>(complex, -1);
+
+        auto complex_input_shape =
+            std::make_shared<v3::ShapeOf>(fw_node_complex_input->input(0).get_source_output(), element::i32);
```
Comment (mvafin) [inline src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp:63]:
```suggestion
            std::make_shared<v3::ShapeOf>(fw_node_complex_input->input_value(0), element::i32);
```

File: src/frontends/pytorch/src/transforms/rfftn_complex_replacer.hpp [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+#include "openvino/pass/graph_rewrite.hpp"
+#include "openvino/pass/pass.hpp"
+
+namespace ov {
+namespace frontend {
+namespace pytorch {
+namespace pass {
+
+class RFFTNComplexReplacer : public ov::pass::MatcherPass {
```
Comment (mvafin) [inline src/frontends/pytorch/src/transforms/rfftn_complex_replacer.hpp:15]: Please add some description for the transformation

File: src/frontends/pytorch/src/transforms/irfftn_complex_replacer.hpp [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+#include "openvino/pass/graph_rewrite.hpp"
+#include "openvino/pass/pass.hpp"
+
+namespace ov {
+namespace frontend {
+namespace pytorch {
+namespace pass {
+
+class IRFFTNComplexReplacer : public ov::pass::MatcherPass {
```
Comment (mvafin) [inline src/frontends/pytorch/src/transforms/irfftn_complex_replacer.hpp:15]: Please add some description

File: src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+            normalized_irfftn = irdft;
+        } else if (norm == "ortho") {
+            auto sqrt_n = std::make_shared<v0::Sqrt>(n);
+            normalized_irfftn = std::make_shared<v1::Multiply>(irdft, sqrt_n);
+        } else {
+            return false;
+        }
+
+        copy_runtime_info({irfftn_op, fw_node_complex_input}, normalized_irfftn);
+        replace_node(irfftn_op, normalized_irfftn);
```
Comment (mvafin) [inline src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp:132]: Please also preserve friendly name of the initial operation
`normalized_irfftn->set_friendly_name(irfftn_op->get_friendly_name());`

File: src/frontends/pytorch/src/transforms/rfftn_complex_replacer.cpp [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+            if (auto real_op = cast_fw_node(out, "aten::real")) {
+                auto squeezed = std::make_shared<v0::Squeeze>(normalized_rfftn_splitted->output(0), const_neg_1);
+                copy_runtime_info({rfftn_op, real_op}, squeezed);
+                replace_node(real_op, squeezed);
+                rval = true;
+            }
+            if (auto imag_op = cast_fw_node(out, "aten::imag")) {
+                auto squeezed = std::make_shared<v0::Squeeze>(normalized_rfftn_splitted->output(1), const_neg_1);
+                copy_runtime_info({rfftn_op, imag_op}, squeezed);
+                replace_node(imag_op, squeezed);
```
Comment (mvafin) [inline src/frontends/pytorch/src/transforms/rfftn_complex_replacer.cpp:121]: Same comments about preserving node names

File: src/frontends/pytorch/src/transforms/rfftn_complex_replacer.cpp [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        }
+
+        std::string norm;
+        if (const auto& fw_node_mode = std::dynamic_pointer_cast<ov::op::util::FrameworkNode>(
+                rfftn_op->input_value(3).get_node_shared_ptr())) {
+            const auto& attrs = fw_node_mode->get_attrs();
+            if (attrs.find("string_value") != attrs.end()) {
+                norm = attrs.at("string_value");
+            } else {
+                norm = "backward";
```
Comment (mvafin) [inline src/frontends/pytorch/src/transforms/rfftn_complex_replacer.cpp:87]: Is this the case when input is none node?

File: src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+        auto input_rank_scalar = std::make_shared<v0::Squeeze>(input_rank);
+
+        bool dim_use_default = is_none_node(irfftn_op->input_value(2).get_node_shared_ptr());
+        bool s_use_default = is_none_node(irfftn_op->input_value(1).get_node_shared_ptr());
+        // Can be None constant, when used check s_use_default.
+        auto raw_s_input_maybe = concat_list_construct(irfftn_op->input_value(1)).get_node_shared_ptr();
+
+        std::shared_ptr<ov::Node> dim;
+        if (!dim_use_default) {
+            auto dim_range = std::make_shared<v0::Range>(const_scalar_0, input_rank_scalar, const_scalar_1);
```
Comment (mvafin) [inline src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp:74]: Please use lates version of `Range` op

**Review by mvafin — CHANGES_REQUESTED** [review]

File: src/frontends/pytorch/src/utils.cpp (hunk 1 above)
Comment (mmikolajcz) [reply src/frontends/pytorch/src/utils.cpp]: Done

File: src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp (hunk 2 above)
Comment (mmikolajcz) [reply src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp]: Done

File: src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp (hunk 3 above)
Comment (mmikolajcz) [reply src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp]: Done

File: src/frontends/pytorch/src/transforms/rfftn_complex_replacer.hpp (hunk 4 above)
Comment (mmikolajcz) [reply src/frontends/pytorch/src/transforms/rfftn_complex_replacer.hpp:15]: Comments were added

File: src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp (hunk 6 above)
Comment (mmikolajcz) [reply src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp:149]: Done

File: src/frontends/pytorch/src/transforms/rfftn_complex_replacer.cpp (hunk 7 above)
Comment (mmikolajcz) [reply src/frontends/pytorch/src/transforms/rfftn_complex_replacer.cpp:141]: Done

File: src/frontends/pytorch/src/transforms/rfftn_complex_replacer.cpp (hunk 8 above)
Comment (mmikolajcz) [reply src/frontends/pytorch/src/transforms/rfftn_complex_replacer.cpp:98]: Yes

**Review by mvafin — APPROVED** [review]

**Review by PiotrKrzem — APPROVED** [review]

**Review by tomdol — CHANGES_REQUESTED** [review]
Since the operators derived from `FFTBase` now accept dynamic element type for their inputs, perhaps it would be worth to review their type prop tests and extend them.

File: src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+        if (const auto& fw_node_mode = std::dynamic_pointer_cast<ov::op::util::FrameworkNode>(
+                irfftn_op->input_value(3).get_node_shared_ptr())) {
+            const auto& attrs = fw_node_mode->get_attrs();
+            if (attrs.find("string_value") != attrs.end()) {
+                norm = attrs.at("string_value");
+            } else {
+                norm = "backward";
+            }
+        } else {
+            return false;
```
Comment (mvafin) [inline src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp:127]: Due to merge of #18015 please add some error message using `add_exception_to_fw_node` in cases when your transformation should've worked, but didn't.

File: src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+        std::shared_ptr<ov::Node> normalized_irfftn;
+        if (norm == "forward") {
+            normalized_irfftn = std::make_shared<v1::Multiply>(irdft, n);
+        } else if (norm == "backward") {
+            normalized_irfftn = irdft;
+        } else if (norm == "ortho") {
+            auto sqrt_n = std::make_shared<v0::Sqrt>(n);
+            normalized_irfftn = std::make_shared<v1::Multiply>(irdft, sqrt_n);
+        } else {
+            return false;
```
Comment (mvafin) [inline src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp:144]: Same comment about error message

File: src/frontends/pytorch/src/transforms/rfftn_complex_replacer.cpp [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+            normalized_rfftn = std::make_shared<v1::Divide>(rdft, n);
+        } else if (norm == "backward") {
+            // No normalization
+            normalized_rfftn = rdft;
+        } else if (norm == "ortho") {
+            // Normalize by 1/sqrt(n)
+            auto sqrt_n = std::make_shared<v0::Sqrt>(n);
+            normalized_rfftn = std::make_shared<v1::Divide>(rdft, sqrt_n);
+        } else {
+            return false;
```
Comment (mvafin) [inline src/frontends/pytorch/src/transforms/rfftn_complex_replacer.cpp:121]: Same comment about error message

**Review by mvafin — COMMENTED** [review]
These comments are not blocking the merge of this PR, but nice to have

File: src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp (hunk 10 above)
Comment (mmikolajcz) [reply src/frontends/pytorch/src/transforms/irfftn_complex_replacer.cpp:128]: Done

File: src/frontends/pytorch/src/transforms/rfftn_complex_replacer.cpp [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+            }
+            if (auto imag_op = cast_fw_node(out, "aten::imag")) {
+                auto squeezed = std::make_shared<v0::Squeeze>(normalized_rfftn_splitted->output(1), const_neg_1);
+                copy_runtime_info({rfftn_op, imag_op}, squeezed);
+                squeezed->set_friendly_name(imag_op->get_friendly_name());
+                replace_node(imag_op, squeezed);
+                rval = true;
+            }
+        }
+        return rval;
```
[the remark this answers is not in the document]
Comment (mmikolajcz) [reply src/frontends/pytorch/src/transforms/rfftn_complex_replacer.cpp:152]: Done

**Pull request merged**